From 81ae33c35f2b2bebe7d88e092f82e73b2a822896 Mon Sep 17 00:00:00 2001
From: Robert Lai <c_rlai@codeaurora.org>
Date: Thu, 10 Sep 2015 17:12:38 -0700
Subject: [PATCH] msm-core: add notification support for cpu power stats ready

Add a new notification for msm-core driver to be able to notify when the
power stats is ready.  The HMP scheduler can take advantage of this to
do pre-calculation of power cost estimation.

CRs-fixed: 849655
Change-Id: I6ee0856e190d9020605a357fb9d00f49f2174ebd
Signed-off-by: Joonwoo Park <joonwoop@codeaurora.org>
---
 drivers/power/qcom/msm-core.c | 12 ++++++++++++
 include/linux/cpu.h           |  1 +
 2 files changed, 13 insertions(+)

diff --git a/drivers/power/qcom/msm-core.c b/drivers/power/qcom/msm-core.c
index 406f097..a20e73f 100644
--- a/drivers/power/qcom/msm-core.c
+++ b/drivers/power/qcom/msm-core.c
@@ -343,6 +343,14 @@ static void clear_static_power(struct cpu_static_info *sp)
 	kfree(sp);
 }
 
+static BLOCKING_NOTIFIER_HEAD(msm_core_stats_notifier_list);
+
+int register_cpu_pwr_stats_ready_notifier(struct notifier_block *nb)
+{
+	return blocking_notifier_chain_register(&msm_core_stats_notifier_list,
+						nb);
+}
+
 static int update_userspace_power(struct sched_params __user *argp)
 {
 	int i;
@@ -418,6 +426,9 @@ static int update_userspace_power(struct sched_params __user *argp)
 			}
 			cpu_stats[cpu].ptable = per_cpu(ptable, cpu);
 			repopulate_stats(cpu);
+
+			blocking_notifier_call_chain(
+				&msm_core_stats_notifier_list, cpu, NULL);
 		}
 	}
 	spin_unlock(&update_lock);
@@ -545,6 +556,7 @@ static int msm_core_stats_init(struct device *dev, int cpu)
 		pstate[i].freq = cpu_node->sp->table[i].frequency;
 
 	per_cpu(ptable, cpu) = pstate;
+
 	return 0;
 }
 
diff --git a/include/linux/cpu.h b/include/linux/cpu.h
index 65d9822..58f5a51 100644
--- a/include/linux/cpu.h
+++ b/include/linux/cpu.h
@@ -279,6 +279,7 @@ static inline void enable_nonboot_cpus(void) {}
 
 struct cpu_pwr_stats *get_cpu_pwr_stats(void);
 void trigger_cpu_pwr_stats_calc(void);
+int register_cpu_pwr_stats_ready_notifier(struct notifier_block *nb);
 
 enum cpuhp_state {
 	CPUHP_OFFLINE,
-- 
1.9.1

